pkg/expr/reader_test.go

```diff
@@ -0,0 +1,143 @@
+package expr
+
+import (
+	"encoding/json"
+	"testing"
+
+	"github.com/grafana/grafana-plugin-sdk-go/data/utils/jsoniter"
+	data "github.com/grafana/grafana-plugin-sdk-go/experimental/apis/data/v0alpha1"
+	"github.com/grafana/grafana/pkg/services/featuremgmt"
+	"github.com/stretchr/testify/require"
+)
+
+func TestReaderReduceMode(t *testing.T) {
+	testData := []struct {
+		name          string
+		bytes         []byte
+		expectedError string
+	}{
+		{
+			name: "no_settings",
+			bytes: []byte(`
+				{
+					"refId": "B",
+					"datasource": {
+						"type": "__expr__",
+						"uid": "__expr__"
+					},
+					"reducer": "last",
+					"expression": "A",
+					"window": "",
+					"type": "reduce"
+				}
+			`),
+			expectedError: "",
+		},
+		{
+			name: "mode_dropnn",
+			bytes: []byte(`
+				{
+					"refId": "B",
+					"datasource": {
+						"type": "__expr__",
+						"uid": "__expr__"
+					},
+					"reducer": "last",
+					"expression": "A",
+					"window": "",
+					"settings": {
+						"mode": "dropNN"
+					},
+					"type": "reduce"
+				}
+			`),
+			expectedError: "",
+		},
+		{
+			name: "mode_replacenn",
+			bytes: []byte(`
+				{
+					"refId": "B",
+					"datasource": {
+						"type": "__expr__",
+						"uid": "__expr__"
+					},
+					"reducer": "last",
+					"expression": "A",
+					"window": "",
+					"settings": {
+						"mode": "replaceNN",
+						"replaceWithValue": 42
+					},
+					"type": "reduce"
+				}
+			`),
+			expectedError: "",
+		},
+		{
+			name: "mode_strict",
+			bytes: []byte(`
+				{
+					"refId": "B",
+					"datasource": {
+						"type": "__expr__",
+						"uid": "__expr__"
+					},
+					"reducer": "last",
+					"expression": "A",
+					"window": "",
+					"settings": {
+						"mode": ""
+					},
+					"type": "reduce"
+				}
+			`),
+			expectedError: "",
+		},
+		{
+			name: "mode_invalid",
+			bytes: []byte(`
+				{
+					"refId": "B",
+					"datasource": {
+						"type": "__expr__",
+						"uid": "__expr__"
+					},
+					"reducer": "last",
+					"expression": "A",
+					"window": "",
+					"settings": {
+						"mode": "invalid-mode"
+					},
+					"type": "reduce"
+				}
+			`),
+			expectedError: "unsupported reduce mode",
+		},
+	}
+
+	for _, test := range testData {
+		t.Run("TestReduceReader:"+test.name, func(t *testing.T) {
+			var q data.DataQuery
+
+			err := json.Unmarshal(test.bytes, &q)
+			require.NoError(t, err)
+
+			raw, err := json.Marshal(q)
+			require.NoError(t, err)
+
+			iter, err := jsoniter.ParseBytes(jsoniter.ConfigDefault, raw)
+			require.NoError(t, err)
+
+			reader := NewExpressionQueryReader(featuremgmt.WithFeatures())
+
+			_, err = reader.ReadQuery(q, iter)
+
+			if test.expectedError == "" {
+				require.NoError(t, err)
+			} else {
+				require.ErrorContains(t, err, test.expectedError)
+			}
```

Comment:
I'm fairly new to Go, so forgive me if this isn't a good suggestion. But I've had a play with it, and I _think_ it's an improvement.

I'm not really a fan of a couple of things here
1. Using `""` to represent a `nil` error. Couldn't `""` be interpreted as an error which contained no error-message? They're different "Classes" of objects - I feel like we shouldn't confuse them
2. I'd rather avoid a code-branch here if we can, and do the same thing every time.

How would you feel about this? Switch the type of `expectedError` from `string` to `error`:
```suggestion
			require.Equal(t, err, test.expectedError)
```

Reply:
i tried to do something similar, but with `error.Is`, see https://github.com/grafana/grafana/pull/94956 
this is how this is done usually, but i'm honestly not sure if this is worth the effort. wdyt?